R/rnet-clean.R

```diff
@@ -80,61 +84,174 @@
+    sfheaders::sf_to_df(line2vertices(rnet))
+  )
+  data.table::setkey(internal_points, x, y)
+  duplicated_internal_points <- internal_points[duplicated(internal_points)]
+
+  # 4 - Split at duplicated internal points
+  if (nrow(duplicated_internal_points) > 0) {
+    message("Splitting rnet object at the duplicated internal points.")
+    rnet <- my_st_split(rnet, duplicated_internal_points)
+  }
+
+  # 5 - Maybe we need to rebuild the tbl_df structure
+  if (rebuild_tbl) {
+    rnet <- sf::st_as_sf(dplyr::as_tibble(rnet))
+  }
+
+  rnet
+}
+
+# Not sure if it should be exported (maybe yes). In any case: ADD DOCS
+# FIXME: ADD DOCS
+my_st_split <- function(rnet, points) {
+  # Extract coordinates from rnet and points
+  # rnet must be an sf data.frame with LINESTRING geometry (no ZM) while points must be an
+  # sfc with MULTIPOINT geometry
+  rnet_coordinates <- data.table::data.table(sfheaders::sf_to_df(rnet))
+
+  # Check Z/M dimensions
+  if (any(c("z", "m") %in% colnames(rnet_coordinates))) {
+    warning("The Z/M dimensions will be lost.", call. = FALSE, immediate. = TRUE)
+  }
+
+  # I cannot split a LINESTRING at a boundary point (since there is nothing "at
+  # the other side" of the boundary point), so I need to find ID of all shared
+  # internal points
+  id_not_boundary <- rnet_coordinates[
+    duplicated(linestring_id) & duplicated(linestring_id, fromLast = TRUE),
+    which = TRUE
+  ]
+  id_shared <- rnet_coordinates[points, on = c("x", "y"), which = TRUE]
+  id <- intersect(id_not_boundary, id_shared)
+
+  # Now I need to duplicate the coordinates of the shared internal points (since
+  # they will become boundary points for the new LINESTRING(s)). I use sort
+  # since I want to preserve the order within the points.
+  rnet_coordinates <- rnet_coordinates[sort(c(1:.N, id))]
+
+  # I will build the new sfc LINESTRING using sfheaders::sfc_linestring so I
+  # need to build an ID for each new LINESTRING. The ID for the new
+  # LINESTRING(s) is created starting from the L1 column in rnet_coordinates and
+  # incrementing the id by 1 at each break point.
+  break_id <- double(nrow(rnet_coordinates))
+  break_id[id + 1:length(id)] <- 1L # The ID is shifted by 1 at each break point
+  break_id <- cumsum(break_id)
+  rnet_coordinates$new_linestring_id <- rnet_coordinates$linestring_id + break_id
+
+  # Now I can create the new LINESTRING sfc
+  new_linestring <- sfheaders::sfc_linestring(
+    obj = rnet_coordinates,
+    x = "x",
+    y = "y",
+    linestring_id = "new_linestring_id"
+  )
+
+  # Exclude Z and/or M dimension
+  attr(new_linestring, "z_range") <- NULL
+  attr(new_linestring, "m_range") <- NULL
+
+  # Determine the ID(s) of the old LINESTRING(s) (i.e. I need to create an sf
+  # object with the new geometry column and the old fields)
+  old_rnet_id <- rnet_coordinates[
+    ,
+    list(linestring_id = unique(linestring_id)),
+    by = new_linestring_id
+  ]$linestring_id
+
+  # Create the new object:
+  sf::st_sf(
```

Comment:
Can this be sped-up by using `sfheaders::sfc_linestring`?

Reply:
I don't know, I will test it later.

Reply:
At the end I decided to use `st_sf` since 1) the difference is minimal since that's not the time consuming step and 2) I still need to use st_as_sf for CRS, precision and so on (that are ignored by sfheaders)